qiskit_machine_learning/circuit/library/qnn_circuit.py

```diff
@@ -0,0 +1,217 @@
+
+    At least one parameter has to be provided. If a feature map and an ansatz is provided, the
+    number of qubits must be the same.
+
+    In case number of qubits is provided along with either a feature map, an ansatz or both, a
+    potential mismatch between the three inputs with respect to the number of qubits is
+    resolved by constructing the ``QNNCircuit`` with the given number of qubits.
+    If one of the ``QNNCircuit`` properties is set after the class construction, a new valid
+    configuration will be derived that considers the latest property update. This ensures that the
+    classes properties are consistent at all times.
+
+    Example:
+
+    .. code-block:: python
+
+        from qiskit_machine_learning.circuit.library import QNNCircuit
+        qnn_qc = QNNCircuit(2)
+        print(qnn_qc)
+        # prints:
+        #      ┌──────────────────────────┐»
+        # q_0: ┤0                         ├»
+        #      │  ZZFeatureMap(x[0],x[1]) │»
+        # q_1: ┤1                         ├»
+        #      └──────────────────────────┘»
+        # «     ┌──────────────────────────────────────────────────────────┐
+        # «q_0: ┤0                                                         ├
+        # «     │  RealAmplitudes(θ[0],θ[1],θ[2],θ[3],θ[4],θ[5],θ[6],θ[7]) │
+        # «q_1: ┤1                                                         ├
+        # «     └──────────────────────────────────────────────────────────┘
+
+        print(qnn_qc.num_qubits)
+        # prints: 2
+
+        print(qnn_qc.input_parameters)
+        # prints: ParameterView([ParameterVectorElement(x[0]), ParameterVectorElement(x[1])])
+
+        print(qnn_qc.weight_parameters)
+        # prints: ParameterView([ParameterVectorElement(θ[0]), ParameterVectorElement(θ[1]),
+        #         ParameterVectorElement(θ[2]), ParameterVectorElement(θ[3]),
+        #         ParameterVectorElement(θ[4]), ParameterVectorElement(θ[5]),
+        #         ParameterVectorElement(θ[6]), ParameterVectorElement(θ[7])])
+    """
+
+    def __init__(
+        self,
+        num_qubits: int | None = None,
+        feature_map: QuantumCircuit | None = None,
+        ansatz: QuantumCircuit | None = None,
+    ) -> None:
+        """
+        Args:
+            num_qubits:  Number of qubits. Optional if feature_map or ansatz is provided, otherwise
+                         required. If not provided num_qubits defaults from the sizes of
+                         feature_map and ansatz.
+            feature_map: A feature map. Optional if num_qubits or ansatz is provided, otherwise
+                         required. If not provided defaults to ``ZZFeatureMap`` or ``ZFeatureMap``
+                         if num_qubits is determined to be 1.
+            ansatz:      An ansatz. Optional if num_qubits or feature_map is provided, otherwise
+                         required. If not provided defaults to ``RealAmplitudes``.
+
```

Comment:
> About knowing the num qubits upfront at creation time. The whole notion of a BlueprintCircuit was it was a template where the num_qubits might not be known upfront and could be later set. Now it can be set later, but cannot be an unknown at the creation point it seems. I know if a FeatureMap instance from what we have today are passed its feature_dimension has to be set (which ends up for what we have so far as same number of qubits). That the Feature Map is a blueprint circuit, where we need to know the feature dim, which ends up defining the num qubits. I guess its comparable.

I don't really get the point here. At least, one of the parameters must be set and the number of qubits is derived from the value of this  parameter. Later, the number of qubits can be updated and, I think, this is the desired behavior.

Reply:
QNNCircuit extends BlueprintCircuit (which is why I mentioned it in second sentence). So QNNCircuit is a BlueprintCircuit. But it does not behave that way like RealAmplitudes(), another Blueprint circuit, which I can create without defining num qubits. Its more like ZZFeatureMap though the parameter there is feature_dimension (which has a correspondance though to num qubits) - hence I mentioned at the end maybe its comparable to that in some respects. As long as things are clearly stated for its behavior (which is the text preceding that to try to summarize things, as for me the class docstring is kinda lengthy in spelling it all out) it seems fine.